### PR TITLE
Milestone 1

### PR DESCRIPTION
**Pull request by scilio**
Contains all changes for milestone 1 of https://forum.grin.mw/t/request-for-funding-scilio-coinswap-implementation/9149?u=scilio

Fee will be included in payload as part of milestone 2.

### CODE REVIEW

File: src/server.rs
Comment (phyro) [inline src/server.rs:18]: nitpick: is this going to be used with `false` in the next milestones?

File: src/ser.rs
Comment (yeastplume) [inline src/ser.rs:1]: Given this is all copied from grin_core, is there a reason why the code here can't just be imported?

**Review by yeastplume — APPROVED** [review]
Caveat: I haven't spent tons of time considering the coinswap proposal and its development.

From an engineering perspective, this all looks like a good start to me and appears to cover phase 1 of the development plan as proposed in the original thread. It appears to present the API as derived in the thread and everything thus far demonstrates a solid understanding of the requirements and how to go about developing them (and we don't need to remind you to write tests covering everything) so glad to have you around @scilio!

This project is in its infancy here, and I'm an advocate of not putting up a load of premature hurdles to early development, based on what I've seen so far I'd be comfortable providing scilio with write access to this repo and letting him run with it for the time being (while trying to keep everything in PRs for review purposes). It will be easier to keep up with changes that way than trying to do a mega-review on all of the proposed phase 2 work, and I wouldn't want progress slowed down waiting for formal reviews on every single change just yet.

I just have one comment for now about some apparently redundant code (there may have been a reason for it,) and I'll let @tromp evaluate the API for the time being, but I'll get into more detailed reviews as we progress. It looks as if the real meat is going to come in the next phase, and I look forward to starting to test-run a few instance of this as the project progresses!

File: src/server.rs
Comment (phyro) [inline src/server.rs:212]: Another one we'll eventually want to add is to check that we accept only 1 swap for a given input - this way we close the attack vector when someone creates two swaps for the same input commitment. Similarly, the last node will want to check that no two swaps create the same output commitment.

File: src/secp.rs
Comment (phyro) [inline src/secp.rs:50]: Not sure if really needed, but when the comsig is implemented, we could write a simple test that asserts the validity of the signature is preserved e.g. `is_valid(deser(ser(sig)), pubkey, msg)` still holds.

**Review by phyro — APPROVED** [review]
I think this is a great start. I've only left a few small notes rather than requests for a change.
I'm a bit weak on the cryptographic side, but I'll try to grasp the implementation concepts better as the projects goes to the next milestones. Thanks for bringing this to life!

File: src/server.rs
Comment (tromp) [reply src/server.rs:212]: Multiple occurrences of the same input or output indicates either malice, a malfunctioning wallet, or multiple wallets sharing the same seed (which can be considered a form of malfunctioning too). In this case we can either drop all but one of the spends, or drop all of them.

Comment (phyro) [reply src/server.rs:212]: I vote for dropping all of them. This way we can see the malfunctioning frequency faster because honest users will complain.

File: README.md
Comment (tromp) [inline README.md:12]: How do these params relate to the data provisioning below? I don't understand why there is a proof of ownership but no rangeproof. And what is data?

File: src/error.rs
Comment (tromp) [inline src/error.rs:48]: There should perhaps be something like a DuplicateInput and DuplicateOutput error.

File: src/server.rs
Comment (scilio) [reply src/server.rs:18]: It should be configured to true for node 1, false for node 2...n
It's not clear yet what ServerConfig will look like in its final form, so that field is likely to change.

File: README.md
Comment (scilio) [reply README.md:12]: `data` refers to the array of encrypted payloads P<sub>i</sub>...P<sub>n</sub>. Maybe `payloads` would be a better name.

File: src/server.rs
Comment (tromp) [reply src/server.rs:18]: I think it might be worthwhile to know that you are server i out of n.
That allows you to check for the correct size of payloads, and whether you're the final node.

**Review by tromp — APPROVED** [review]
Looks good so far; I've made some minor comments regarding swap API params, duplicate inputs/outputs, and representing mixnode indices.

File: README.md
Comment (tromp) [reply README.md:12]: I had concluded that data was too short (at a mere 34 bytes) to contain a rangeproof:-(
The other fields seem to have realistic lengths.
But I agree, payloads is more descriptive.
What is the role of msg?

File: src/ser.rs
Comment (scilio) [reply src/ser.rs:1]: grin_core functions as a library, but hasn't yet matured into a stable, broadly-consumable one. I was concerned we'd end up chasing frequent api changes. But alas, my investigations for milestone 2 have shown it's unavoidable. I will be replacing this with grin_core imports as part of that milestone.

File: src/server.rs
Comment (scilio) [reply src/server.rs:212]: Wallets sharing the same seed is probably too common for us to just drop all duplicate inputs. We should decide between the earliest submission or the most recent one.

Duplicate outputs does seem more like a form of maliciousness or wallet malfunction. For that, it may be better to drop both. node n wouldn't know which one was submitted first anyway.

File: README.md
Comment (scilio) [reply README.md:12]: True, I should've included a rangeproof in the payload. I will change to `payloads` and replace this example with a real one for 3 nodes. We'll see 2 of these small payloads followed by a larger one with the rangeproof included.

I wasn't sure what message should be signed for the ComSig, so I just added a msg field. I guess the Onion field could be signed instead. Thoughts?

File: src/error.rs
Comment (scilio) [reply src/error.rs:48]: Good idea. I'll include those errors when I write the duplicate checking logic in milestone 2

File: src/server.rs
Comment (scilio) [reply src/server.rs:18]: Good suggestion. Milestone 3 is where ServerConfig will really come into play. I will make note of this for when I get to milestone 3.

File: src/server.rs
Comment (tromp) [reply src/server.rs:212]: If you keep one of two inputs sent from two seed-sharing wallets, then the wallet whose input you dropped will malfunction anyway when its input moves in a way it doesn't understand. So I'm inclined to agree with phyro, that it's better to drop all.

File: README.md
Comment (tromp) [reply README.md:12]: I'm fine with signing with the onion field as message.

File: src/server.rs
Comment (yeastplume) [reply src/server.rs:212]: When wallets restore from a seed they have no way of knowing what the last used derivation index was (they can guess from the last output belonging to it, but it's obviously not 100%), so non-malicious duplicate outputs can occur with non-zero frequency. The wallet stores an output's MMR position (once the output is found on-chain) to disambiguate it internally in this instance. It might be possible to make MMR position an optional field here, so duplicate outputs can be dropped if they're the same and neither has an MMR position, of if they both have the same MMR position. If only one provides an MMR pos it is preferred and other dropped. (note, I need to check whether wallet stores mmr position or insertion index, I think it's position)

File: src/ser.rs
Comment (yeastplume) [reply src/ser.rs:1]: Fair enough, I thought you'd have a reason. grin_core isn't very well presented as a library at the moment, but API changes should be very infrequent. We need to change CI to trigger builds of downstream packages when anything in core changes (wallet and this package), so we'll cover that eventuality there.

File: src/server.rs
Comment (tromp) [reply src/server.rs:212]: what scenario do you envision in which a user is hampered by duplicate invalidation?
does it involve a user mixing within one day both before and after restore?
btw, i'd rather not complicate the protocol with optional fields.

Comment (phyro) [reply src/server.rs:212]: Suppose we have `n` wallets sharing the same seed `w1..wn`. Only one wallet will create the output, so perhaps we could say that if the output was created with `wi`, then it's up to `wi` to create the coinswap. A wallet can know whether it created an output or not unless there has been a restore, but if a restore was done, you can just create a coinswap after 25 hours or so.

**Review by j01tz — APPROVED** [review]
This PR seems to implement the basics as described in the [first milestone](https://forum.grin.mw/t/request-for-funding-scilio-coinswap-implementation/9149). Beyond that, I don't have much to add that hasn't already been brought up in these comments. 

I do have some comments about the specification itself but will keep them in the [forum thread](https://forum.grin.mw/t/mimblewimble-coinswap-proposal/8322). 

It's nice to have you here @scilio, looking forward to seeing your next work.

**Comment by phyro** [comment]
I've opened an issue https://github.com/mimblewimble/mwixnet/issues/2 to track the minor leftovers from this discussion so we don't forget them in milestone 2.

**Pull request merged**
